pipeline/pipeline.go

```diff
@@ -487,24 +530,27 @@ func (p *Pipeline) maintenance() {
 		p.antispamer.maintenance()
 		p.metricsHolder.maintenance()
 
-		totalCommitted := p.totalCommitted.Load()
-		deltaCommitted := int(totalCommitted - lastCommitted)
-
-		totalSize := p.totalSize.Load()
-		deltaSize := int(totalSize - lastSize)
+		// set new values and get the diffs
+		deltaInputEvents := inputEvents.updateValue(p.inputEvents.Load())
+		deltaInputSize := inputSize.updateValue(p.inputSize.Load())
+		deltaOutputEvents := outputEvents.updateValue(p.outputEvents.Load())
+		deltaOutputSize := outputSize.updateValue(p.outputSize.Load())
 
-		rate := int(float64(deltaCommitted) * float64(time.Second) / float64(interval))
-		rateMb := float64(deltaSize) * float64(time.Second) / float64(interval) / 1024 / 1024
-
-		tc := totalCommitted
-		if totalCommitted == 0 {
-			tc = 1
-		}
+		stats.GetCounter(p.subsystemName(), inputEventsCountMetric).Add(deltaInputEvents)
```

Comment:
Can the delta be less than 0?

```
func (c *counter) Add(v float64) {
	if v < 0 {
		panic(errors.New("counter cannot decrease in value"))
	}
```

Reply:
i added a check to the `update` function of the deltawrapper, we'll log an error if `diff < 0` and return abs(diff).